iModelJsNodeAddon/JsCloudSqlite.cpp

```diff
@@ -256,7 +274,7 @@ struct JsCloudContainer : CloudContainer, Napi::ObjectWrap<JsCloudContainer> {
     }
 
     Napi::Value UploadChanges(NapiInfoCR info) {
-        RequireWriteLock();

```

Comment:
Maybe everywhere we call RequireWriteLock, we should be calling CheckWriteLockHeldByCurrentUser()? Or potentially replace requriewritelock with your implementation?

Reply:
Resolved in commit [commit 35f01fd017118730a226782e7d4cbebc0752b487](https://github.com/iTwin/imodel-native/pull/819/commits/35f01fd017118730a226782e7d4cbebc0752b487)

Reply:
Thanks, are there other places in the code where we potentially rely on m_writeLockHeld and perhaps we shouldn't be checking that?

Reply:
There's only one more place that is using m_writeLockHeld, `void Connect(NapiInfoCR info)`. I think that one is fine according to the comments
![image](https://github.com/user-attachments/assets/38438893-1415-435f-8981-097725e52968)

Reply:
what if we connected our container, grabbed the write lock, disconnected our container, let user2 grab the write lock, then connected our container again?

I imagine we'd end up in the catch within ResumeWriteLock, but I do wonder if m_writeLockHeld could still be true in this case?

Reply:
Since none of these actions call (AbandonChanges | ReleaseWriteLock | ClearWriteLock) so m_writeLockHeld will remain true in this case. I will write a test to validate this

Reply:
User1 and User2 are in different processes so they don't share the same m_writeLockHeld. User2's m_writeLockHeld is false if user2 didn't call acquireWriteLock(). User1's m_writeLockHeld is still true in this case.

However, since User2 has grabbed the write lock, we want User1 to AbandonChanges() when User1 reconnect, right?

Reply:
Yes, correct

Reply:
Resolved in [commit c32f13881e48297481c3903552c2ed6109a6a921](https://github.com/iTwin/imodel-native/pull/819/commits/c32f13881e48297481c3903552c2ed6109a6a921)